### PR TITLE
Base classes for the relationship algorithms + Never-Together implemenation.

### PR DESCRIPTION
**Pull request by Bohne9**
This PR implements a base class for the implementaitons of the log-skeleton relationship algorithms.
Further the never-together algorithm is implemented.

### CODE REVIEW

File: src/components/logic/relationship.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        # Each apply_to_trace returns a list of results.
+        #   results might look like -> [[(a, b), (b, c)], [(b, d)], [(c, a)]]
+        # Flattening this example leads to: [(a, b), (b, c), (b, d), (c, a)]
+        # flattenedResult = [val for traceRes in results for val in traceRes]
+
+        return results
+
+    def apply_to_trace(self, trace, a1, a2) -> bool:
+        """Apply the matching algorithm to each pair of activities."""
+        if self.activity_pair_matches(trace, a1, a2):
```
Comment (AveraGE0) [inline src/components/logic/relationship.py:111]: Maybe I am missing something but doesnt activity_pair_matches only take 2 parameters in the current implementation?

Comment (Bohne9) [reply src/components/logic/relationship.py:111]: I forgot to change it in the abstract class. 😅 The implementation in the subclass takes 3 parameters (that is how it is suppossed to be).
I'll change that real quick! Thank you

Comment (AveraGE0) [reply src/components/logic/relationship.py:109]: Ok great! I'll approve it once its changed.

**Review by AveraGE0 — APPROVED** [review]

**Comment by Bohne9** [comment]
[Implement base classes to implement the LS relationship algorithms.](https://trello.com/c/JJcBUPHH/44-implement-base-classes-to-implement-the-ls-relationship-algorithms)

**Comment by Bohne9** [comment]
[Implement a class for the never-together-relation algorithm.](https://trello.com/c/FYvRdUWC/50-implement-a-class-for-the-never-together-relation-algorithm)

**Pull request merged**